Nx extension2

wenyingd opened the pull request:
Add support for Nicira extension, and also add support for several standard OF match and actions.
This pull request could be leveraged to help generate openflow entries for OKN implementation.

wenyingd committed:
Nicira extension: Add support for Nicira extended match

Add support to match Nicira OF extension field: regX[m..n], ct_state,
ct_mark, ct_zone, conj_id
Add a general functions to find Match field using OXX field name, and
this function supports to find match field with or without mask

Signed-off-by: wenyingd <wenyingd@vmware.com>

wenyingd committed:
Add support for standard OF match of ARP addresses

1. Define messages for ARP Protocol Address and ARP Hardware Address, and
use these two messages for match arp_sha, arp_tha, arp_spa, arp_tpa.
2. Add code to support generating match with key NXM_OF_ARP_SPA,
NXM_OF_ARP_TPA, NXM_NX_ARP_SHA, NXM_NX_ARP_THA

Signed-off-by: wenyingd <wenyingd@vmware.com>

wenyingd committed:
Add support for standard OF match of STCP port

Add support for match stcp_src and stcp_dst port

Signed-off-by: wenyingd <wenyingd@vmware.com>

wenyingd committed:
Nicira extension: Add support for Nicira extended action

Add support for Nicira OF extension actions: conjunction, ct, load, move,
resubmit, nat, output_reg,dec_ttl,dec_ttl_cnt_ids
For these actions: load, move, output_reg, a match field is needed as a
valid field, so function `FindFieldHeaderByName` in nx_util.go would be
invoked to search target.

Signed-off-by: wenyingd <wenyingd@vmware.com>

edwardbadboy commented on openflow13/nx_match.go:
```diff
@@ -0,0 +1,269 @@
+
+// ct_state = -new
+func (s *CTStates) UnsetNew() {
+	s.data |= 0 << NX_CT_STATE_NEW_OFS
```
To unset a bit without affecting other bits, maybe a more reliable way is like

```s.data = s.data &^ (1 << NX_CT_STATE_NEW_OFS)```

wenyingd replied:
Greate catch! I would update.

wenyingd committed:
Bug fix: Unset ct_state

Use Operator '&^' on target state flag when client wants to unset
the flag.

Signed-off-by: wenyingd <wenyingd@vmware.com>

edwardbadboy commented on openflow13/action.go:
```diff
@@ -3,7 +3,6 @@ package openflow13
 import (
 	"encoding/binary"
 	"errors"
-
```
Seems we can keep the blank line. The golang import can be organized in groups. https://github.com/golang/go/wiki/CodeReviewComments#imports

wenyingd replied:
Done.

wenyingd replied:
Thanks, I would update in the latest update.

edwardbadboy commented on openflow13/action.go:
```diff
@@ -94,16 +93,87 @@ func DecodeAction(data []byte) Action {
 		a = new(ActionHeader)
+	case ActionType_Experimenter:
+		// For Experimenter message, there should be at least 8 bytes following action type field,
+		// including experimenter_id(4 byte), vendor(2 byte), and subtype(2 byte)
```
So is it actually like:
```
type (2 bytes), length (2 bytes)
experimenter_id (4 bytes), vendor (2 bytes), subtype (2 bytes)
```

I see that `subtype := binary.BigEndian.Uint16(data[8:])`, so here subtype covers actually vendor+subtype in the above comment?

wenyingd replied:
Nope, binary.BigEndian.Uint16(data[8:]) would parse a uint16 number from 8th bit of data, and it would only consume 2 bytes, it is descided by Uint16.

antoninbas commented on openflow13/action.go:
```diff
@@ -94,16 +94,87 @@ func DecodeAction(data []byte) Action {
+			return nil, errors.New("The byte array has wrong size for openflow experimenter message")
+		}
+		v := binary.BigEndian.Uint32(data[4:8])
+		if v == NxExperimenterID {
```
I suggest moving this if block to a `DecodeNxAction` function

wenyingd replied:
Sure, I would move it to a specific function.

antoninbas commented on openflow13/action.go:
```diff
@@ -94,16 +94,87 @@ func DecodeAction(data []byte) Action {
+			case NXAST_RESUBMIT_TABLE:
+				a = new(NXActionResubmitTable)
+			case NXAST_OUTPUT_REG:
+				a = newNXActionOutputReg()
```
does this need to be `newNXActionOutputReg()`, why not `new(NXActionOutputReg)`? Would the data not get unmarshalled properly?

wenyingd replied:
new(NXActionOutputReg) should be OK, I would try to update and re-test.

antoninbas commented on openflow13/action.go:
```diff
@@ -94,16 +94,87 @@ func DecodeAction(data []byte) Action {
+				a = newNXActionOutputReg()
+			case NXAST_LEARN:
+			case NXAST_EXIT:
+			case NXAST_DEC_TTL:
```
Potentially a stupid question: I see a lot of actions here (`NXAST_SET_QUEUE`, `NXAST_DEC_TTL` `NXAST_PUSH_MPLS`) that are available in OpenFlow 1.1+ (and in particular in OpenFlow 1.3 which is the only version that this library claims to support); what is the point of including these actions?

wenyingd replied:
OpenFlow 1.1+ should be the earliest version announced to support these actions. This library supports Openflow1.3, this means this library should be compatible with all match/actions defined in versions 1.0 ~ 1.3. Besides, some OVS extensions are not leveraged in OKN code, so I only add implementations for those are needed by OKN in this library, that's why you might see several actions are empty.

antoninbas replied:
Thanks for the clarification. I didn't know this library was supposed to support versions earlier than 1.3.

antoninbas commented on openflow13/match.go:
```diff
@@ -174,6 +175,33 @@ func (m *MatchField) UnmarshalBinary(data []byte) error {
+func (m *MatchField) UnmarshalHeader(data []byte) error {
+	var err error
+	if len(data) < int(4) {
+		err = fmt.Errorf("the byte array has wrong size to unmarshal MatchField header")
```
suggestion: "the byte array is not large enough to unmarshal MatchField header"

(this suggestion applies to several other locations as well)

wenyingd replied:
Thanks for giving suggestion for error. I have quick question, existent code use these two template to describe byte length check error, should I follow them or leverage OKN fromats?

https://github.com/wenyingd/libOpenflow/blob/master/protocol/ethernet.go#L162
https://github.com/wenyingd/libOpenflow/blob/master/openflow13/action.go#L247

antoninbas replied:
The first link is fine: "The []byte is too short to unmarshal a full VLAN header.". The second one is not a correct sentence.

antoninbas commented on openflow13/match.go:
```diff
@@ -174,6 +175,33 @@ func (m *MatchField) UnmarshalBinary(data []byte) error {
+	m.Class = binary.BigEndian.Uint16(data[n:])
+	n += 2
+	fieldWithMask := data[n]
+	m.HasMask = bool(fieldWithMask&1 == 1)
```
I don't think the `bool` cast is needed

wenyingd replied:
I would change.

antoninbas commented on openflow13/nx_action.go:
```diff
@@ -0,0 +1,956 @@
+
+func (a *NXActionHeader) MarshalBinary() (data []byte, err error) {
+	data = make([]byte, a.Len())
+	b := make([]byte, 0)
```
that doesn't seem necessary to me (same comment for other locations below as well), do you think it helps with readability?

wenyingd replied:
I don't get your point. Do you mean `b` is not needed? As I need to Marshal NxActionHeader firstly, and a byte slice is needed in this operation. If I don't use `b`, I don't know how to copy header bytes into the message bytes.

antoninbas replied:
I don't understand why you have:
```
b := make([]byte, 0)
b, err = a.ActionHeader.MarshalBinary()
```
instead of simply:
```
b, err := a.ActionHeader.MarshalBinary()
```
Am I missing something?

antoninbas commented on openflow13/nx_action.go:
```diff
@@ -0,0 +1,956 @@
+	a.Type = Type_Experimenter
+	a.Length = a.NXActionHeader.Len() + 6
+	a.InPort = inPort
+	a.pad = make([]byte, 3, 3)
```
There is no point in providing a capacity value if it is equal to the size (couple places below as well).

wenyingd replied:
Done.

antoninbas commented on openflow13/nx_action.go:
```diff
@@ -0,0 +1,956 @@
+	// padding the message if the length is not an integer multiple of 8 bytes
+	if n < int(a.Length) {
+		var padN = int(a.Length) - n
+		pad := make([]byte, padN, padN)
```
no need to specify capacity

wenyingd replied:
Done.

antoninbas commented on openflow13/nx_action.go:
```diff
@@ -0,0 +1,956 @@
+	if n < int(a.Length) {
+		var padN = int(a.Length) - n
+		pad := make([]byte, padN, padN)
+		copy(optData, pad)
```
I may be missing something but what is this `copy` for?

wenyingd replied:
This copy could be removed, I would update.

antoninbas commented on openflow13/nx_action.go:
```diff
@@ -0,0 +1,956 @@
+	n += optN
+
+	// padding the message if the length is not an integer multiple of 8 bytes
+	if n < int(a.Length) {
```
I am not sure I understand the point of this TBH: if `data` already has the correct length , there is nothing you need to do. You already have the necessary padding.

wenyingd replied:
If padding is the last part, it might be right, but for some messages, padding is in the middle, I think padding should be needed. Or do you think it will work if I could only move index of byte array if padding is in the middle? Anyway I could try to remove this operation of "copy padding".

antoninbas replied:
The new version looks better to me. Thanks.

antoninbas commented on openflow13/nx_action.go:
```diff
@@ -0,0 +1,956 @@
+}
+
+func (a *NXActionCTNAT) Len() (n uint16) {
+	a.Length = uint16(math.Ceil(float64(a.Length)/float64(8))) * 8
```
I think the following is more concise / elegant:
```
a.Length = ((a.Length + 7) / 8) * 8
```

wenyingd replied:
Done.

antoninbas commented on openflow13/nx_action.go:
```diff
@@ -0,0 +1,956 @@
+}
+
+func (a *NXActionCTNAT) MarshalBinary() (data []byte, err error) {
+	optData := make([]byte, 0)
```
I am not sure you need `optData` at all. Couldn't you move the option handling to the end?

wenyingd replied:
Do you mean I should copy the bytes of each option to data directly? I will change.

antoninbas commented on openflow13/nx_action.go:
```diff
@@ -0,0 +1,956 @@
+	return a.Length
+}
+
+func (a *NXActionCTNAT) MarshalBinary() (data []byte, err error) {
```
It may be worth it to add a unit test for this action

wenyingd replied:
I have added integration test cases for leveraging this action in ofnet, and then I could ensure if or not these actions could be realized. Maybe I could add some unit tests in this repo.

antoninbas replied:
I see that you added a unit test. Thanks, I think it was necessary.

antoninbas commented on openflow13/nx_action.go:
```diff
@@ -0,0 +1,956 @@
+
+func (a *NXActionCTNAT) SetSNAT() error {
+	if a.Flags&(^uint16(NX_NAT_F_MASK)) != 0 || a.Flags&NX_NAT_F_DST != 0 {
+		return errors.New("SNAT action should be exclusively with DNAT")
```
that's not correct english
suggestion: "the SNAT and DNAT actions are mutually exclusive"

(applies to multiple locations below as well)

wenyingd replied:
Done.

antoninbas commented on openflow13/nx_action.go:
```diff
@@ -0,0 +1,956 @@
+}
+
+func (a *NXActionCTNAT) SetDNAT() error {
+	if a.Flags&NX_NAT_F_SRC != 0 {
```
you do not check `NX_NAT_F_MASK` like in `SetSNAT`, is this normal?

wenyingd replied:
Check with `NX_NAT_F_MASK` is not needed here, I would remove from SetSNAT too.

antoninbas commented on openflow13/nx_action.go:
```diff
@@ -0,0 +1,956 @@
+	OfsNbits uint16      // (ofs << 6 | (Nbits -1)
+	SrcField *MatchField // source nxm_nx_reg
+	MaxLen   uint16      // Max length to send to controller if chosen port is OFPP_CONTROLLER
+	zero     []uint8     // 6 uint8 with all Value as 0, reserved
```
I think for such a case, a Go array may be simpler (you can use default initialization) and more efficient: `zeros [6]uint8`. It also makes it clear that there are 6 uint8 values.

wenyingd replied:
Done.

antoninbas commented on openflow13/nx_util.go:
```diff
@@ -0,0 +1,278 @@
+}
+
+func FindFieldHeaderByName(fieldName string, hasMask bool) (*MatchField, error) {
+	var err = fmt.Errorf("failed to find header by name %s", fieldName)
```
what's the rationale for declaring the error at the beginning of the function?

wenyingd replied:
I found some code in this repo has this style to declare error at the beginning, so I want to follow it. I could move it to the line it should be invoked.

antoninbas replied:
Leaving it as it is is fine by me. I find it strange but I don't have a string opinion.

antoninbas commented on openflow13/nxext_test.go:
```diff
@@ -0,0 +1,215 @@
+	action1 := NewNXActionResubmit(portID)
+	data1, err := action1.MarshalBinary()
+	if err != nil {
+		t.Errorf("Failed to invoke NXAST_RESUBMIT_TABLE.MarshalBinary: %v", err)
```
should be a `Fatalf` IMO. If the call fails then `data1` is not correct and there is no point in calling `UnmarshalBinary` on it after.

This probably applies to other locations as well.

wenyingd replied:
Sure，I would update.

antoninbas replied:
I do think there are more places where it should be `Fatalf` and not `Errorf`. Basically you have to ask yourself: does it make sense to proceed with the test if this fails, or will everything else just fail automatically because of this one failure.

antoninbas commented on openflow13/nxext_test.go:
```diff
@@ -0,0 +1,215 @@
+	binary.BigEndian.PutUint32(data, headerInt)
+	_ = testMFHeader.UnmarshalHeader(data)
+	if testMFHeader.Class != tgtField.Class {
+		t.Errorf("Failed to UnmarshalHeader: class")
```
that's a weird error message: you didn't fail to unmarshal, the result is just not correct

 Did you mean something like "Unmarshalled header has incorrect 'Class' field"?

wenyingd replied:
Yes, you are correct. I would change as you suggested.

antoninbas commented on openflow13/nxext_test.go:
```diff
@@ -0,0 +1,215 @@
+	testMFHeader := new(MatchField)
+	var data []byte = make([]byte, 4)
+	binary.BigEndian.PutUint32(data, headerInt)
+	_ = testMFHeader.UnmarshalHeader(data)
```
I think you should check the returned error, why not?

wenyingd replied:
I would add error check here.

wenyingd committed:
Revise error messages and DecodeAction

Extract code for decoding NX actions into a single function.

Signed-off-by: wenyingd <wenyingd@vmware.com>

antoninbas commented on openflow13/nx_action.go:
```diff
@@ -929,18 +983,18 @@ func (a *NXActionDecTTL) UnmarshalBinary(data []byte) error {
 	a := &NXActionDecTTL{
 		NXActionHeader: NewNxActionHeader(NXAST_DEC_TTL),
-		zeros:          []uint8{0, 0, 0, 0},
+		zeros:          [4]uint8{},
```
I don't think this needed

wenyingd replied:
Do you mean field zeros is not needed? If so, I'm afraid the message encoding might not be as expected.

antoninbas replied:
That's just me nitpicking. I mean that the explicit initialization is not needed, since `zeros` is now a fixed size array. By default all elements of the array will be initialized to 0

antoninbas commented on openflow13/nx_action.go:
```diff
@@ -949,7 +1003,7 @@ func NewNXActionDecTTL() *NXActionDecTTL {
 	a := &NXActionDecTTL{
 		NXActionHeader: NewNxActionHeader(NXAST_DEC_TTL_CNT_IDS),
-		zeros:          []uint8{0, 0, 0, 0},
+		zeros:          [4]uint8{},
```
Same as above

antoninbas commented on openflow13/nxext_test.go:
```diff
@@ -14,7 +15,7 @@ func TestNXActionResubmit(t *testing.T) {
 	action1 := NewNXActionResubmit(portID)
```
I believe this test would benefit from using Go subtests (https://golang.org/pkg/testing/#hdr-Subtests_and_Sub_benchmarks) to isolate the different test actions from each other. Even better if you can make the test table-driven as well.

antoninbas commented on openflow13/nxext_test.go:
```diff
@@ -14,7 +15,7 @@ func TestNXActionResubmit(t *testing.T) {
-		t.Errorf("Failed to invoke NXAST_RESUBMIT_TABLE.MarshalBinary: %v", err)
+		t.Fatalf("Failed to invoke NXAST_RESUBMIT_TABLE.MarshalBinary: %v", err)
 	}
 	testAction1 := new(NXActionResubmitTable)
```
I am a bit confused, why isn't this `newNXActionResubmitTable` like for all subsequent cases?

wenyingd replied:
OVS use resubmit for three types of messages, all these messages have similar message struct but different action subtype.  So the unit test is expecting to cover encoding/decoding for all these messages.

antoninbas reviewed (commented):
Thanks for addressing my comments. I added a few more as I was going through the new diff.

blp commented on openflow13/match.go:
```diff
@@ -333,45 +441,61 @@ const (
+	NXM_NX_TUN_METADAT4  = 44  /* nicira extension: tun_metadata, for Geneve header variable data */
+	NXM_NX_TUN_METADAT5  = 45  /* nicira extension: tun_metadata, for Geneve header variable data */
+	NXM_NX_TUN_METADAT6  = 46  /* nicira extension: tun_metadata, for Geneve header variable data */
+	NXM_NX_TUN_METADAT7  = 47  /* nicira extension: tun_metadata, for Geneve header variable data */
```
It is unusual to leave off the "A". I would spell out TUN_METADATA, including the A.

Also, the comments on these variables differ slightly in format. I would make them consistent.

wenyingd replied:
Done.

blp commented on openflow13/nx_util.go:
```diff
@@ -0,0 +1,278 @@
+	return &MatchField{Class: class, Field: field, Length: fieldLength, HasMask: false}
+}
+
+// oxxFieldHeaderMap is map to find target field header without mask using an OVS known OXX field name
```
I think you mean OXM instead of OXX.

blp commented on openflow13/nx_util.go:
```diff
@@ -0,0 +1,278 @@
+	"OXM_OF_ARP_SPA":        newMatchFieldHeader(OXM_CLASS_OPENFLOW_BASIC, OXM_FIELD_ARP_SPA, 4),
+	"OXM_OF_ARP_TPA":        newMatchFieldHeader(OXM_CLASS_OPENFLOW_BASIC, OXM_FIELD_ARP_TPA, 4),
+	"OXM_OF_ARP_SHA":        newMatchFieldHeader(OXM_CLASS_OPENFLOW_BASIC, OXM_FIELD_ARP_SHA, 6),
+	"OXM_OF_ARP_THA":        newMatchFieldHeader(OXM_CLASS_OPENFLOW_BASIC, OXM_FIELD_ARP_SHA, 6),
```
Should be OXM_FIELD_ARP_THA not ..._SHA.

wenyingd replied:
Done.

blp commented on openflow13/nx_util.go:
```diff
@@ -0,0 +1,278 @@
+	"OXM_OF_IPV6_EXTHDR":    newMatchFieldHeader(OXM_CLASS_OPENFLOW_BASIC, OXM_FIELD_IPV6_EXTHDR, 2),
+}
+
+// oxxFieldHeaderWildMap is map to find target field header with mask using an OVS known OXX field name
```
OXM, not OXX.

blp commented on openflow13/nx_util.go:
```diff
@@ -0,0 +1,278 @@
+	return ofs<<6 | (nBits - 1)
+}
+
+func encodeOfs(ofsNbits uint16) uint16 {
```
I think this is a decode function, not an encode one.

wenyingd replied:
Done.

blp commented on openflow13/nx_util.go:
```diff
@@ -0,0 +1,278 @@
+	return ofsNbits >> 6
+}
+
+func decodeNbits(ofsBnits uint16) uint16 {
```
Parameter name looks like a typo for ofsNbits

blp commented on openflow13/match.go:
```diff
@@ -1323,3 +1330,100 @@ func NewTunnelIpv4DstField(tunnelIpDst net.IP, tunnelIpDstMask *net.IP) *MatchFi
+	return nil
+}
+
+func NewArpThaField(arpTha net.HardwareAddr) *MatchField {
```
I think that this function and NewArpShaField() are the same code except for the OXM_FIELD_ARP_* constant. I would suggest refactoring them to use a common helper function.

blp commented on openflow13/match.go:
```diff
@@ -1323,3 +1330,100 @@ func NewTunnelIpv4DstField(tunnelIpDst net.IP, tunnelIpDstMask *net.IP) *MatchFi
+	return nil
+}
+
+func NewArpTpaField(arpTpa net.IP) *MatchField {
```
I think this is the same as the next function. I suggest refactoring in terms of a new helper.

wenyingd replied:
As existent code in libOpenflow is using this style, every field is constructed in a specific function, and all code is similar. So I just follow the style. I think it could be optimized in another MR if necessary.

blp commented on openflow13/nx_match.go:
```diff
@@ -267,3 +267,67 @@ func NewConjIDMatchField(conjID uint32) *MatchField {
+	} else {
+		field, _ = FindFieldHeaderByName("NXM_NX_ARP_SHA", false)
+	}
+
```
I suspect that this can be written as:

> field, _ = FindFieldHeaderByName("NXM_NX_ARP_SHA", mask != nil);

wenyingd replied:
Done.

blp commented on openflow13/nx_match.go:
```diff
@@ -267,3 +267,67 @@ func NewConjIDMatchField(conjID uint32) *MatchField {
 	return field
 }
+
+func NewNxARPShaMatchField(addr net.HardwareAddr, mask net.HardwareAddr) *MatchField {
```
I believe that this and the following function can be refactored in terms of a common helper.

blp commented on openflow13/match.go:
```diff
@@ -1331,6 +1331,36 @@ func NewTunnelIpv4DstField(tunnelIpDst net.IP, tunnelIpDstMask *net.IP) *MatchFi
 }
 
+// SCTP_DST field
+func NewSctpDstField(port uint16) *MatchField {
```
I think that these functions can be written in terms of a common helper.

blp commented on openflow13/action.go:
```diff
@@ -101,9 +101,80 @@ func DecodeAction(data []byte) Action {
+			case NXAST_OUTPUT_REG2:
+				a = new(NXActionOutputReg)
+			case NXAST_REG_LOAD2:
+			case NXAST_CNJUNCTION:
```
I am not sure why the "O" is left out of CONJUNCTION.

wenyingd replied:
Done.

blp commented on openflow13/nx_action.go:
```diff
@@ -897,3 +897,60 @@ func (a *NXActionOutputReg2) UnmarshalBinary(data []byte) error {
+	*NXActionHeader
+	controllers uint16  // number of controller
+	zeros       []uint8 // 4 byte with zeros
+}
```
Maybe this is a misinterpretation of NXAST_DEC_TTL_CNT_IDS. This actions has variable length: the fixed data above is followed by a sequence of controller IDs, the number specified by 'controllers' in the fixed data. I don't see anything about variable-length data in this implementation.

wenyingd replied:
I add new struct specific for NXAST_DEC_TTL_CNT_IDS, and add a field for controller IDs.

wenyingd committed:
Bugfix: Add support for ONFOXM_ET_TCP_FLAGS

OVS use ONFOXM_ET_TCP_FLAGS(42) in OF1.3, the match message class is
0xffffffff, and ONF_Experimenter_ID(0x4f4e4600) is added before the
actual value and mask.

This patch could fix errors in `DecodeMatchField` while decoding
`tcp_flags` messages sent from OVS.

Signed-off-by: wenyingd <wenyingd@vmware.com>

wenyingd committed:
Correct spelling errors

Signed-off-by: wenyingd <wenyingd@vmware.com>

wenyingd commented:
Updates already merged in upstreaming repo, so close this PR.
